modules/packages/ParallelIO.chpl

```diff
@@ -0,0 +1,705 @@
+  only a header, and then a continuous stream of deserializeable values of the
+  same type).
+
+  There are distributed and local versions of each procedure. The distributed
+  procedures return a block distributed array over the provided locales, while
+  the local procedures return a default rectangular array.
+
+  * :proc:`readParallelLines`: read each of the lines of a file as a ``string``
+    or ``bytes`` value
+  * :proc:`readParallelDelimited`: read a file where each value is strictly
+    separated by a delimiter, and the delimiter cannot be found in the
+    value (e.g., CSV)
+  * :proc:`readParallel`: read a file where each value is separated by delimiter
+
+  There are also some helper procedures that can be used to implement other
+  parallel I/O procedures:
+
+  * :proc:`findDelimChunks`: find a set of byte offsets that divide a file into
+    roughly equal chunks where each chunk begins with a delimiter
+  * :proc:`findDelimChunksChecked`: find a set of byte offsets that divide a file
+    into roughly equal chunks where each chunk begins with a delimiter and each
+    chunk starts with a deserializable value
+  * :proc:`findItemOffsets`: get a prefix sum of the number of items in each
+    chunk of a file, where the chunks are defined by the ``byteOffsets`` array,
+    and each item is separated by the given delimiter
+*/
+@unstable("the 'parallelIO' module is unstable and subject to change in a future release")
+module ParallelIO {
+  private use IO, BlockDist, List, CTypes, OS;
+
+  /*
+    Read a file's lines in parallel into a block-distributed array.
+
+    This routine is similar to :proc:`readParallelDelimited`, except that it
+    reads each line as a :type:`~String.string` or :type:`~Bytes.bytes` value.
+
+    :arg filePath: a path to the file to read from
+    :arg lineType: which type to represent a line: either ``string`` or ``bytes``
+    :arg nTasks: the number of tasks to use per locale
+        (if ``-1``, query ``here.maxTaskPar`` on each locale)
+    :arg header: how to handle the file header (see :record:`headerPolicy`)
+    :arg targetLocales: the locales to read the file on
+
+    :returns: a block-distributed array of ``lineType`` values
+
+    :throws: :class:`OffsetNotFoundError` if a starting offset cannot be found
+             in any of the chunks
+
+    See :proc:`~IO.open` for other errors that could be thrown when attempting
+  */
+  proc readParallelLines(filePath: string, type lineType = string, header = headerPolicy.noHeader,
+                         nTasks: int = -1, targetLocales: [?d] locale = Locales
+  ): [] lineType throws
+    where lineType == string || lineType == bytes
+  {
+    param delim = b"\n";
+
+    const fMeta = open(filePath, ioMode.r),
+          fileBounds = 0..<fMeta.size,
+          byteOffsets = findDelimChunks(fMeta, delim, d.size, fileBounds, header),
+          itemOffsets = findItemOffsets(fMeta, delim, byteOffsets);
```

Comment:
I see that you are doing this ahead-of-time but it requires more passes over the file than not doing it this way.

The alternative I see is this:
 * compute totalBytes = fileSize - headerSize
 * compute taskChunk = totalBytes / numTasks
 * task `i` is responsible for reading any line that begin in `headerSize+i*taskChunk` up to but not including `headerSize+(i+1)*taskChunk`
   * task 0 needs to read the 0th line, always
   * other tasks start reading at `headerSize+i*taskChunk` but initially they are just looking for a newline. Once they find a newline, they start reading delimited elements.

(and it can work similarly for other delimited data).

Is the issue with this approach that, while it balances the amount to read across threads, it does not balance the number of items? In other words, if we read lines into a Block distributed array, there could be a mismatch between this file-offset-based idea of responsibility and the item-index one that Block will use?

Perhaps the approach I am describing would work well for a parallel `file.lines` but not as well for a routine that creates a Block array from the lines?

Reply:
I went with this design because I wanted to be able to preallocate the `results` array and read items directly into it, meaning that I need to know the total number of items in the file (and within each chunk) before I start reading them. This could be more work than it's worth, but I do like that it results in a nearly even number of items per locale in the block case.

The `readParallel` and `readParallelLocal` procedures do something more similar to what you're describing because there is no efficient way to precompute the number of items in that case.

I agree that the approach you lined out could also be more efficient for a lines iterator.

Reply:
No so sure I agree that readParallel does what I'm describing. The key point of the thing I'm describing is that there is no reading (other than header processing, potentially) before the tasks start working on their regions. And then, each task reads contiguously, but might read slightly outside of its region.